[2.9 ACRA] crash adding audio to card

https://couchdb.ankidroid.org/acralyzer/_design/acralyzer/index.html#/report-details/9a93b88a-ca83-4870-b023-555ca95c5290

```
java.lang.RuntimeException: Failure delivering result ResultInfo{who=null, request=1, result=-1, data=Intent { dat=content://com.android.providers.media.documents/document/audio:36799 flg=0x1 }} to activity {com.ichi2.anki/com.ichi2.anki.multimediacard.activity.MultimediaEditFieldActivity}: java.lang.ArrayIndexOutOfBoundsException: length=1; index=1
at android.app.ActivityThread.deliverResults(ActivityThread.java:4491)
at android.app.ActivityThread.handleSendResult(ActivityThread.java:4534)
at android.app.ActivityThread.-wrap20(Unknown Source:0)
at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1752)
at android.os.Handler.dispatchMessage(Handler.java:105)
at android.os.Looper.loop(Looper.java:164)
at android.app.ActivityThread.main(ActivityThread.java:6942)
at java.lang.reflect.Method.invoke(Native Method)
at com.android.internal.os.Zygote$MethodAndArgsCaller.run(Zygote.java:327)
at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1374)
Caused by: java.lang.ArrayIndexOutOfBoundsException: length=1; index=1
at com.ichi2.anki.multimediacard.fields.BasicAudioClipFieldController.onActivityResult(BasicAudioClipFieldController.java:9)
at com.ichi2.anki.multimediacard.activity.MultimediaEditFieldActivity.onActivityResult(MultimediaEditFieldActivity.java:3)
at android.app.Activity.dispatchActivityResult(Activity.java:7556)
at android.app.ActivityThread.deliverResults(ActivityThread.java:4487)
... 9 more
```

This code can also crash like so:

https://couchdb.ankidroid.org/acralyzer/_design/acralyzer/index.html#/report-details/ff975412-a839-460c-8d93-ce5fda5b834d

```
java.lang.RuntimeException: Failure delivering result ResultInfo{who=null, request=1, result=-1, data=Intent { dat=file:///storage/emulated/0/Download/pronunciation_cs_červená.mp3 }} to activity {com.ichi2.anki/com.ichi2.anki.multimediacard.activity.MultimediaEditFieldActivity}: java.lang.NullPointerException: Attempt to invoke interface method 'boolean android.database.Cursor.moveToFirst()' on a null object reference
at android.app.ActivityThread.deliverResults(ActivityThread.java)
at android.app.ActivityThread.handleSendResult(ActivityThread.java)
at android.app.ActivityThread.access$1300(ActivityThread.java)
at android.app.ActivityThread$H.handleMessage(ActivityThread.java)
at android.os.Handler.dispatchMessage(Handler.java)
at android.os.Looper.loop(Looper.java)
at android.app.ActivityThread.main(ActivityThread.java)
at java.lang.reflect.Method.invoke(Native Method)
at java.lang.reflect.Method.invoke(Method.java:372)
at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java)
at com.android.internal.os.ZygoteInit.main(ZygoteInit.java)
Caused by: java.lang.NullPointerException: Attempt to invoke interface method 'boolean android.database.Cursor.moveToFirst()' on a null object reference
at com.ichi2.anki.multimediacard.fields.BasicAudioClipFieldController.onActivityResult(BasicAudioClipFieldController.java:5)
at com.ichi2.anki.multimediacard.activity.MultimediaEditFieldActivity.onActivityResult(MultimediaEditFieldActivity.java:3)
at android.app.Activity.dispatchActivityResult(Activity.java)
```

--- a/AnkiDroid/src/main/java/com/ichi2/anki/ModelFieldEditor.java
+++ b/AnkiDroid/src/main/java/com/ichi2/anki/ModelFieldEditor.java
@@ -182,9 +182,9 @@ public class ModelFieldEditor extends AnkiActivity {
                             .replaceAll("[\'\"\\n\\r\\[\\]\\(\\)]", "");
 
                     if (fieldName.length() == 0) {
-                        showToast(getResources().getString(R.string.toast_empty_name));
+                        UIUtils.showThemedToast(this, getResources().getString(R.string.toast_empty_name), true);
                     } else if (containsField(fieldName)) {
-                        showToast(getResources().getString(R.string.toast_duplicate_field));
+                        UIUtils.showThemedToast(this, getResources().getString(R.string.toast_duplicate_field), true);
                     } else {
                         //Name is valid, now field is added
                         try {
@@ -235,7 +235,7 @@ public class ModelFieldEditor extends AnkiActivity {
 
 
         if (mFieldLabels.size() < 2) {
-            showToast(getResources().getString(R.string.toast_last_field));
+            UIUtils.showThemedToast(this, getResources().getString(R.string.toast_last_field), true);
         } else {
             try {
                 mCol.modSchema();
@@ -282,9 +282,9 @@ public class ModelFieldEditor extends AnkiActivity {
                         String fieldLabel = mFieldNameInput.getText().toString()
                                 .replaceAll("[\'\"\\n\\r\\[\\]\\(\\)]", "");
                         if (fieldLabel.length() == 0) {
-                            showToast(getResources().getString(R.string.toast_empty_name));
+                            UIUtils.showThemedToast(this, getResources().getString(R.string.toast_empty_name), true);
                         } else if (containsField(fieldLabel)) {
-                            showToast(getResources().getString(R.string.toast_duplicate_field));
+                            UIUtils.showThemedToast(this, getResources().getString(R.string.toast_duplicate_field), true);
                         } else {
                             //Field is valid, now rename
                             try {
@@ -332,12 +332,12 @@ public class ModelFieldEditor extends AnkiActivity {
                         try {
                             pos = Integer.parseInt(newPosition);
                         } catch (NumberFormatException n) {
-                            showToast(getResources().getString(R.string.toast_out_of_range));
+                            UIUtils.showThemedToast(this, getResources().getString(R.string.toast_out_of_range), true);
                             return;
                         }
 
                         if (pos < 1 || pos > mFieldLabels.size()) {
-                            showToast(getResources().getString(R.string.toast_out_of_range));
+                            UIUtils.showThemedToast(this, getResources().getString(R.string.toast_out_of_range), true);
                         } else {
                             // Input is valid, now attempt to modify
                             try {
@@ -461,13 +461,6 @@ public class ModelFieldEditor extends AnkiActivity {
             }
         }
         return false;
-    }
-
-
-    private void showToast(CharSequence text) {
-        int duration = Toast.LENGTH_SHORT;
-        Toast toast = Toast.makeText(this, text, duration);
-        toast.show();
     }
 
 

--- a/AnkiDroid/src/main/java/com/ichi2/anki/multimediacard/fields/BasicAudioClipFieldController.java
+++ b/AnkiDroid/src/main/java/com/ichi2/anki/multimediacard/fields/BasicAudioClipFieldController.java
@@ -1,3 +1,18 @@
+/****************************************************************************************
+ * Copyright (c) 2020 Mike Hardy <github@mikehardy.net>                                 *
+ *                                                                                      *
+ * This program is free software; you can redistribute it and/or modify it under        *
+ * the terms of the GNU General Public License as published by the Free Software        *
+ * Foundation; either version 3 of the License, or (at your option) any later           *
+ * version.                                                                             *
+ *                                                                                      *
+ * This program is distributed in the hope that it will be useful, but WITHOUT ANY      *
+ * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A      *
+ * PARTICULAR PURPOSE. See the GNU General Public License for more details.             *
+ *                                                                                      *
+ * You should have received a copy of the GNU General Public License along with         *
+ * this program.  If not, see <http://www.gnu.org/licenses/>.                           *
+ ****************************************************************************************/
 package com.ichi2.anki.multimediacard.fields;
 
 import android.app.Activity;

--- a/AnkiDroid/src/main/java/com/ichi2/anki/multimediacard/fields/BasicAudioClipFieldController.java
+++ b/AnkiDroid/src/main/java/com/ichi2/anki/multimediacard/fields/BasicAudioClipFieldController.java
@@ -87,15 +87,31 @@ public class BasicAudioClipFieldController extends FieldControllerBase implement
         if ((resultCode != Activity.RESULT_CANCELED) && (requestCode == ACTIVITY_SELECT_AUDIO_CLIP)) {
 
             Uri selectedClip = data.getData();
-            Timber.d(selectedClip.toString());
 
             // Get information about the selected document
-            String[] queryColumns = { MediaStore.MediaColumns.DISPLAY_NAME, MediaStore.MediaColumns.SIZE };
+            String[] queryColumns = { MediaStore.MediaColumns.DISPLAY_NAME, MediaStore.MediaColumns.SIZE, MediaStore.MediaColumns.MIME_TYPE };
             Cursor cursor = mActivity.getContentResolver().query(selectedClip, queryColumns, null, null, null);
+
+            if (cursor == null) {
+                 UIUtils.showThemedToast(AnkiDroidApp.getInstance().getApplicationContext(),
+                         AnkiDroidApp.getInstance().getString(R.string.multimedia_editor_something_wrong), true);
+                return;
+            }
+
             cursor.moveToFirst();
+            Timber.d("got name/size/type: %s/%s/%s", cursor.getString(0), cursor.getLong(1), cursor.getString(2));
             String audioClipFullName = cursor.getString(0);
-            // Note this could be unsafe, null name, or name without standard extension etc, go off mime types instead?
             String[] audioClipFullNameParts = audioClipFullName.split("\\.");
+            if (audioClipFullNameParts.length < 2) {
+                try {
+                    Timber.d("Audio clip name does not have extension, using second half of mime type");
+                    audioClipFullNameParts = new String[] {audioClipFullName, cursor.getString(2).split("\\/")[1]};
+                } catch (Exception e) {
+                    UIUtils.showThemedToast(AnkiDroidApp.getInstance().getApplicationContext(),
+                            AnkiDroidApp.getInstance().getString(R.string.multimedia_editor_something_wrong), true);
+                    return;
+                }
+            }
             cursor.close();
 
             // We may receive documents we can't access directly, we have to copy to a temp file

--- a/AnkiDroid/src/main/java/com/ichi2/anki/multimediacard/fields/BasicAudioClipFieldController.java
+++ b/AnkiDroid/src/main/java/com/ichi2/anki/multimediacard/fields/BasicAudioClipFieldController.java
@@ -28,16 +28,16 @@ import android.widget.LinearLayout;
 import android.widget.Button;
 import android.widget.TextView;
 
+import com.ichi2.anki.AnkiDroidApp;
 import com.ichi2.anki.CollectionHelper;
 import com.ichi2.anki.R;
+import com.ichi2.anki.UIUtils;
+import com.ichi2.compat.CompatHelper;
 import com.ichi2.libanki.Collection;
 
 import java.io.File;
-import java.io.FileNotFoundException;
-import java.io.FileOutputStream;
 import java.io.IOException;
 import java.io.InputStream;
-import java.io.OutputStream;
 
 import timber.log.Timber;
 
@@ -99,65 +99,42 @@ public class BasicAudioClipFieldController extends FieldControllerBase implement
             cursor.close();
 
             // We may receive documents we can't access directly, we have to copy to a temp file
-            File clipCopy = null;
+            File clipCopy;
             try {
                 clipCopy = File.createTempFile("ankidroid_audioclip_" + audioClipFullNameParts[0],
                         "." + audioClipFullNameParts[1],
                         storingDirectory);
-            } catch (IOException e) {
+                Timber.d("audio clip picker file path is: %s", clipCopy.getAbsolutePath());
+            } catch (Exception e) {
                 Timber.e(e, "Could not create temporary audio file. ");
+                UIUtils.showThemedToast(AnkiDroidApp.getInstance().getApplicationContext(),
+                        AnkiDroidApp.getInstance().getString(R.string.multimedia_editor_something_wrong), true);
+                return;
             }
 
             // Copy file contents into new temp file. Possibly check file size first and warn if large?
             InputStream inputStream = null;
             try {
                 inputStream = mActivity.getContentResolver().openInputStream(selectedClip);
+                CompatHelper.getCompat().copyFile(inputStream, clipCopy.getAbsolutePath());
 
-                // copy the picked file contents to internal media iteratively using Buffered Streams?
-                OutputStream outputStream = null;
-                try {
-                    // FIXME no no no no
-                    clipCopy.setWritable(true, false);
-                    outputStream = new FileOutputStream(clipCopy);
-                    byte buffer[] = new byte[1024];
-                    int length = 0;
-
-                    while((length = inputStream.read(buffer)) > 0) {
-                        outputStream.write(buffer,0,length);
-                    }
-                }
-                catch (IOException e) {
-                    Timber.e(e, "error in creating a file");
-                }
-                finally {
-                    try {
-                        outputStream.close();
-                    }
-                    catch (IOException ioe) {
-                        // nothing?
-                    }
-                }
-
+                // If everything worked, hand off the information
                 mField.setHasTemporaryMedia(true);
-            }
-            catch (FileNotFoundException fnfe) {
-                Timber.e(fnfe, "Unable to open selected audio clip");
+                mField.setAudioPath(clipCopy.getAbsolutePath());
+                mTvAudioClip.setText(mField.getFormattedValue());
+                mTvAudioClip.setVisibility(View.VISIBLE);
+            } catch (Exception e) {
+                Timber.e(e, "Unable to copy audio file from ContentProvider");
+                UIUtils.showThemedToast(AnkiDroidApp.getInstance().getApplicationContext(),
+                        AnkiDroidApp.getInstance().getString(R.string.multimedia_editor_something_wrong), true);
             }
             finally {
-                // close the file descriptor here?
                 try {
                     inputStream.close();
-                }
-                catch (IOException ioe) {
+                } catch (IOException ioe) {
                     // nothing
                 }
             }
-
-            Timber.d("audio clip picker file path is: %s", clipCopy.getAbsolutePath());
-            mField.setAudioPath(clipCopy.getAbsolutePath());
-
-            mTvAudioClip.setText(mField.getFormattedValue());
-            mTvAudioClip.setVisibility(View.VISIBLE);
         }
     }
 

